python_modules/dagster/dagster/core/asset_defs/asset_selection.py

```diff
@@ -0,0 +1,142 @@
+from functools import reduce
+from typing import AbstractSet, FrozenSet, Optional
+
+from typing_extensions import TypeAlias
+
+import dagster._check as check
+from dagster.core.asset_defs.assets import AssetsDefinition
+from dagster.core.selector.subset_selector import (
+    Direction,
+    fetch_connected_assets_definitions,
+    generate_asset_dep_graph,
+    generate_asset_name_to_definition_map,
+)
+
+AssetSet: TypeAlias = AbstractSet[AssetsDefinition]  # makes sigs more readable
+
+
+class AssetSelection(ABC):
+    @staticmethod
+    def all() -> "AllAssetSelection":
+        return AllAssetSelection()
+
+    @staticmethod
+    def keys(*key_strs: str) -> "KeysAssetSelection":
+        return KeysAssetSelection(*key_strs)
+
+    @staticmethod
+    def groups(*group_strs) -> "GroupsAssetSelection":
+        return GroupsAssetSelection(*group_strs)
+
+    def downstream(self, depth: Optional[int] = None) -> "DownstreamAssetSelection":
+        return DownstreamAssetSelection(self, depth=depth)
+
+    def upstream(self, depth: Optional[int] = None) -> "UpstreamAssetSelection":
+        return UpstreamAssetSelection(self, depth=depth)
+
+    def __or__(self, other: "AssetSelection") -> "OrAssetSelection":
+        return OrAssetSelection(self, other)
+
+    def __and__(self, other: "AssetSelection") -> "AndAssetSelection":
+        return AndAssetSelection(self, other)
+
+    def resolve(self, all_assets: AssetSet) -> AssetSet:
+        return Resolver(all_assets).resolve(self)
+
+
+class AllAssetSelection(AssetSelection):
+    pass
+
+
+class AndAssetSelection(AssetSelection):
+    def __init__(self, child_1: AssetSelection, child_2: AssetSelection):
+        self.children = (child_1, child_2)
+
+
+class DownstreamAssetSelection(AssetSelection):
+    def __init__(self, child: AssetSelection, *, depth: Optional[int] = None):
+        self.children = (child,)
+        self.depth = depth
+
+
+class GroupsAssetSelection(AssetSelection):
+    def __init__(self, *children: str):
+        self.children = children
+
+
+class KeysAssetSelection(AssetSelection):
+    def __init__(self, *children: str):
+        self.children = children
+
+
+class OrAssetSelection(AssetSelection):
+    def __init__(self, child_1: AssetSelection, child_2: AssetSelection):
+        self.children = (child_1, child_2)
+
+
+class UpstreamAssetSelection(AssetSelection):
+    def __init__(self, child: AssetSelection, *, depth: Optional[int] = None):
+        self.children = (child,)
+        self.depth = depth
+
+
+# ########################
+# ##### RESOLUTION
+# ########################
+
+
+class Resolver:
+    def __init__(self, all_assets: AssetSet):
+        self.all_assets = all_assets
+        self.asset_dep_graph = generate_asset_dep_graph(list(all_assets))
+        self.all_assets_by_name = generate_asset_name_to_definition_map(all_assets)
+
+    def resolve(self, node: AssetSelection) -> AssetSet:
+        if isinstance(node, AllAssetSelection):
```

Comment:
Did you investigate whether it would make sense to push this down into an abstract method of `AssetSelection` instead of doing these `isinstance` checks?

Reply:
That was the original implementation, but then I converted it to a class for the reason above-- because intermediate data structures are needed over the course of a resolution. It is cleaner to retain these as instance variables of a class then it is to recursively pass them down when for many individual node resolutions they are not needed.